### PR TITLE
Better temp telemetry

### PR DESCRIPTION
**Pull request by nkrackow**
Implements min/max/average temperature of last telemetry period.
Seems to work nicely. 
Printing the counter in `.process()` yields 503 for some channels, 504 for the others and each individual channel roughly alternates. Consistent with 503.5 Hz sample rate.
Average temperature looks very clean now for a 10K ref resistor.

### CODE REVIEW

File: src/temperature_tele.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        let average = self.accumulator / self.counter as f32;
+        let temp = Temperature {
+            average,
+            max: self.max,
+            min: self.min,
+        };
+        self.counter = 0;
+        self.accumulator = 0.;
+        self.max = 0.;
+        self.min = 0.;
```
Comment (jordens) [inline src/temperature_tele.rs:49]:
```suggestion
        self.max = f32::NEGATIVE_INFINITY;
        self.max = f32::INFINITY;
```

File: src/temperature_tele.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+/// Temperature buffer for computing min/max/average of the last telemetry period.
+#[derive(Default, Copy, Clone, Debug)]
+pub struct TemperatureBuffer {
+    min: f32,
+    max: f32,
+    accumulator: f32,
+    counter: u32,
+}
```
Comment (jordens) [inline src/temperature_tele.rs:20]: `stddev` is also very useful, but should probably be done in a separate PR, as you'll first need to look up the right algo.

File: src/temperature_tele.rs (hunk 1 above)
Comment (nkrackow) [reply src/temperature_tele.rs:49]: Ah, obviously...

File: src/temperature_tele.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
+//! # Thermostat_EEM temperature telemetry processing
+
+use serde::Serialize;
+
+/// Temperature telemetry struct.
+#[derive(Serialize, Copy, Clone, Default, Debug)]
+pub struct Temperature {
+    average: f32,
```
Comment (jordens) [inline src/temperature_tele.rs:8]:
```suggestion
    mean: f32,
```

File: src/temperature_tele.rs (hunk 2 above)
Comment (nkrackow) [reply src/temperature_tele.rs:20]: ack.

Comment (nkrackow) [reply src/temperature_tele.rs:20]: Do you have a resource for this?

File: src/temperature_tele.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    accumulator: f32,
+    counter: u32,
+}
+
+impl TemperatureBuffer {
+    /// Add a new temperature sample to the buffer. This will add it to the accumulator,
+    /// update min/max and increment the counter.
+    pub fn add(&mut self, temp: f32) {
+        self.counter += 1;
+        self.accumulator += temp;
```
Comment (jordens) [inline src/temperature_tele.rs]: Does `f32` have enough dynamic range? I suspect it doesn't.
At least make a note that this is OK as long as either the ADC values are sufficiently noisy or the telemetry period is short enough. Otherwise you'll get cancellation effects (addition).

File: src/temperature_tele.rs (hunk 2 above)
Comment (nkrackow) [reply src/temperature_tele.rs:20]: Is [this](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#:~:text=to%20a%20degree.-,Welford%27s%20online%20algorithm,-%5Bedit%5D) the one you have in mind?

Comment (jordens) [reply src/temperature_tele.rs:20]: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Na%C3%AFve_algorithm

The naive one with f64 should be good enough. Or Welford with f32, but maybe costlier due to the division.

File: src/temperature_tele.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        self.max = self.max.max(temp);
+        self.min = self.min.min(temp);
+    }
+
+    /// Process temperature buffer. Calculates average, returns the finalized Temperature type
+    /// and resets the buffer.
+    pub fn process(&mut self) -> Temperature {
```
[the remark this answers is not in the document]
Comment (jordens) [reply src/temperature_tele.rs:34]: And preferrably make both `Buffer` and `Statistics` generic over a certain set of float traits (from `num_traits`).
My guess is that there should be some existing crate for what is in this file.

File: src/temperature_tele.rs (hunk 4 above)
Comment (nkrackow) [reply src/temperature_tele.rs]: I thought about it while implementing it but then forgot at the end. 
Very rough estimate: 25°C * 500 is already 5 digits, so any further sample addition will only be accurate to about 0.01 °C (about 7 digits total for 24 bit mantissa), right?

Comment (jordens) [reply src/temperature_tele.rs:43]: Worst case is t_tele/t_sample*T_max*f32::eps ~ 10*1000*100*1e-7 K = 0.1 K. So yes, it's bad.

File: src/temperature_tele.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jordens) [reply src/temperature_tele.rs]: But the generic thing can also be done later in another PR.

File: src/temperature_tele.rs (hunk 4 above)
Comment (jordens) [reply src/temperature_tele.rs]: And it's even worse for stddev when using the naive algo.

File: src/statistics.rs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    /// Process temperature buffer. Calculates mean, returns the finalized Statistics type
+    /// and resets the buffer.
+    fn from(mut buff: Buffer) -> Statistics {
+        let mean = buff.accumulator / buff.counter as f64;
+        let stat = Statistics {
+            mean: mean as f32,
+            max: buff.max,
+            min: buff.min,
+        };
+        buff.reset();
```
Comment (jordens) [inline src/statistics.rs:23]: No point in resetting it. You are consuming `buff`.

File: src/main.rs [hunk 7]
```diff
@@ -236,6 +240,13 @@ mod app {
             telemetry.monitor.overtemp = gpio.overtemp();
             telemetry.monitor.poe = gpio.poe();
         });
+        // finalize temperature telemetry
+        for ch in InputChannel::into_enum_iter() {
+            telemetry.statistics[ch as usize] = c
+                .shared
+                .ch_statistics_buff
+                .lock(|buff| buff[ch as usize].into());
```
Comment (jordens) [inline src/main.rs:248]: Because `Buffer` is `Copy`, you are (AFAICT) not resetting here. `impl Default` and assign that (see below).

```rust
lock(|buff| {
    let stat = buff[ch as usize].into();
    buff[ch as usize] = Buffer::default();
    stat
}
```

File: src/statistics.rs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+/// Statistics telemetry struct.
+#[derive(Serialize, Copy, Clone, Default, Debug)]
+pub struct Statistics {
+    mean: f32,
+    min: f32,
+    max: f32,
+}
+
+impl From<Buffer> for Statistics {
+    /// Process temperature buffer. Calculates mean, returns the finalized Statistics type
```
Comment (jordens) [inline src/statistics.rs:14]: `From` should not modify the input. Either it consumes it (then modifying is pointless) or it gets it by immutable reference (then it can't).

**Review by jordens — COMMENTED** [review]
Very cool. Something similar would be great for `fls` and `stabilizer`.

File: src/statistics.rs (hunk 8 above)
Comment (nkrackow) [reply src/statistics.rs:14]: I actually committed the same error twice today, earlier with `mut self` instead of `&mut self`..
Easy to just forget.

**Review by jordens — APPROVED** [review]
